ddtrace/internal/processor/stats.py

```diff
@@ -0,0 +1,239 @@
+from ..logger import get_logger
+from ..periodic import PeriodicService
+from ..writer import _human_size
+
+
+if typing.TYPE_CHECKING:
+    from typing import DefaultDict
+    from typing import List
+    from typing import Optional
+
+    from ddtrace import Span
+
+    from ..agent import ConnectionType
+
+
+log = get_logger(__name__)
+
+
+def _is_top_level(span):
+    # type: (Span) -> bool
+    """Return whether the span is a "top level" span."""
+    return (span._local_root is span) or (
+        span._parent is not None and span._parent.service != span.service and span.service is not None
+    )
+
+
+def _is_measured(span):
+    # type: (Span) -> bool
+    """Return whether the span is flagged to be measured or not."""
+    return SPAN_MEASURED_KEY in span.metrics
+
+
+"""
+To aggregate metrics for spans they need to be "uniquely" identified (as
+best as possible). This enables the compression of stat points.
+
+Aggregation can be done using primary and secondary attributes from the span
+stored in a tuple which is hashable in Python.
+"""
+SpanAggrKey = typing.Tuple[
+    str,  # name
+    str,  # service
+    str,  # resource
+    str,  # type
+    int,  # http status code
+    bool,  # synthetics request
+]
+
+
+class SpanAggrStats(object):
+    """Aggregated span statistics."""
+
+    def __init__(self):
+        self.hits = 0
+        self.top_level_hits = 0
+        self.errors = 0
+        self.duration = 0
+        self.ok_distribution = LogCollapsingLowestDenseDDSketch(0.00775, bin_limit=2048)
+        self.err_distribution = LogCollapsingLowestDenseDDSketch(0.00775, bin_limit=2048)
+
+
+def _span_aggr_key(span):
+    # type: (Span) -> SpanAggrKey
+    """Return a hashable key that can be used to uniquely refer to the "same"
+    span.
+    """
+    service = span.service or "unnamed_service"
+    if len(service) > 100:
+        service = service[:100]
+    resource = span.resource or span.name
+    _type = span.span_type
+    if _type and len(_type) > 100:
+        _type = _type[:100]
+
+    status_code = span.metrics.get("http.status_code", None)
+    synthetics = span.context.dd_origin == "synthetics"  # TODO: verify this is the right variant
+    return span.name, service, resource, _type, status_code, synthetics
+
+
+class SpanStatsProcessor(PeriodicService, SpanProcessor):
+    def __init__(self, agent_url, interval=10.0, timeout=1.0):
+        # type: (str, float, float) -> None
+        super(SpanStatsProcessor, self).__init__(interval=interval)
+        self.start()
+        self._agent_url = agent_url
+        self._timeout = timeout
+        self._bucket_size_ns = int(10 * 1e9)  # type: int
+        self._buckets = defaultdict(
+            lambda: defaultdict(SpanAggrStats)
+        )  # type: DefaultDict[int, DefaultDict[SpanAggrKey, SpanAggrStats]]
+        self._endpoint = "/v0.6/stats"
+        self._connection = None  # type: Optional[ConnectionType]
+        self._lock = threading.Lock()
```

Comment:
do we forksafe lock?

Reply:
from conversation: in theory no since on fork we re-initialize the writer and span processors.

Reply:
Yeah so the conclusion here:

The need for a forksafe lock would be to avoid deadlocks. Deadlocks occur when the process forks while a lock is acquired by a thread and since the thread is not carried into the child process the lock is forever locked.

There is the chance that this can happen in the following situation:

If the StatsProcessor acquires the lock and the process forks then the lock will be permanently locked in the child. This is fine assuming we reset the processor before `on_span_finish` is called but the fork check we do in the library is done on span start, not finish.

Thus it's possible that in the child process, before a span is started, a span is finished which would result in a deadlock.

So we might actually want to use a forksafe lock here!